### PR TITLE
llama_sampling_sample with default args is more naively usable 

### PR DESCRIPTION
**Pull request by TheFlipbook**
* Hopefully reduces chances of error case cited in https://github.com/ggerganov/llama.cpp/issues/6475
* Batches populated by either `llama_batch_get_one` or `llama_batch_add` work with default args
  * Previously `get_one` could use the default argument
  * Previously `add` should usually have used the last index where `batch->logits[idx] == true`
* This hopefully encourages the use of `llama_batch_add`
  * By giving expected results when using default arguments.
* Believed to work with any currently well behaved program
  * Default arg now works for both cases (previously would give strange results for add case)
  * Any non-negative number is unaffected and behaves as previously
  * Negative arguments were previously invalid.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -15540,6 +15540,26 @@ float * llama_get_logits_ith(struct llama_context * ctx, int32_t i) {
...
+        return 0;
+    }
+
+    // non-trivial case scan for the last output.
+    for (int32_t i = ctx->output_ids.size() - 1; i >= 0; --i) {
+        const int32_t candidate = ctx->output_ids[i];
+        if (candidate >= 0) {
+            return i;
+        }
+    }
```
Comment (compilade) [inline llama.cpp:15557]: I'm pretty sure there's a faster and simpler way to do this. If `ctx->n_outputs` is set to the total number of outputs after a batch is processed, the last logits would be available at `ctx->logits[ctx->n_outputs - 1]`.

(this could be a special case of `llama_get_logits_ith` instead, to avoid trying to find the last logits index, which is still non-trivial)

In `llama_decode_internal`, the number of outputs is already calculated as `n_outputs`, so a simple `lctx.n_outputs = n_outputs` near the end of the function (way after the compute graph and its inputs are built, outside of the `ubatch` loop) would do the trick.

(BTW, I worked on adding `ctx->output_ids`, `ctx->n_outputs`, and other related stuff in #6122)

Comment (TheFlipbook) [reply llama.cpp:15557]: That seems reasonable to me!

I had been worried about changing the public api of `llama_get_logits_ith`, but if that is viable, then I concur, it seems guaranteed optimal to do it there, rather than on the index.

I had been hopeful that the worst case was uncommon, as it seemed likely that the last index would be set for wanting logits, and therefore the for loop would, in practice, usually exit on the first loop.

Would you like me to update this PR to conform to the suggested code?  I think the suggestion looks great, and even adds a cool piece of functionality of python-style negative indexing.

Comment (compilade) [reply llama.cpp:15557]: > I had been worried about changing the public api of `llama_get_logits_ith`

Since this seems like a backward-compatible change, there is no need to worry. API changes are quite common in `llama.cpp`, there's even a "Recent API changes" section in the `README.md` for this purpose.

> I had been hopeful that the worst case was uncommon, as it seemed likely that the last index would be set for wanting logits, and therefore the for loop would, in practice, usually exit on the first loop.

You are right, the worst case isn't common (at least when using `batch.logits`). Another way would have been to always return `ctx->cparams.n_batch - 1` or `ctx->output_ids.size() - 1`, and let `llama_get_logits_ith` do its asserts to fail in unexpected cases, but this would not work with `llama_batch_get_one`.

Thinking a bit more about this, the worst case is actually quite common in the sense that it happens all the time when using `llama_batch_get_one`, because the "last valid index" in this case is 0.

> Would you like me to update this PR to conform to the suggested code?

Yes. You had the idea of making this, so feel free to make commits in your name.

I hereby allow @TheFlipbook to use my code review suggestion from this comment thread and do absolutely anything with it, including but not limited to committing it (even with modifications) without attributing me.

Comment (TheFlipbook) [reply llama.cpp]: Updated the PR with these suggestions.   Apologies for the double-push, the first one had a formatting mistake.

Let me know if you'd suggest any further updates!  Thanks for all the help!

**Comment by TheFlipbook** [comment]
It looks like `CI / macOS-latest-cmake-x64` might be failing but I'm not quite sure how to verify if the PR is causal.

I'm trying to read the logs, and I see:
`Test #21: test-backend-ops ....................***Failed    0.33 sec`
 and above that lines like this:
 ```
 21:   SQR(type=f32,ne=[10,10,10,10]): ggml_backend_alloc_ctx_tensors_from_buft: tensor sent_0 is too large to fit in a Metal buffer (tensor size: 4096, max buffer size: 0)
21: failed to allocate tensors
```

If I'm following the chain for that error correctly it's because `ggml_backend_metal_buffer_type_get_max_size` is returning 0.  Based on https://developer.apple.com/documentation/metal/mtldevice/2966563-maxbufferlength, I'm not quite sure how this PR could be causal for that error.

Perhaps the CI error is a transient error?

**Review by ggerganov — APPROVED** [review]
The macOS workflow has been failing occasionally recently - last time I looked into this, it seemed the runner can end up randomly without a GPU device for some reason. Anyway, it's not relevant for this change
